autopush_rs/src/util/megaphone.rs

```diff
@@ -0,0 +1,249 @@
+        }
+        changelist
+    }
+
+    /// Add a new service to the ServiceChangeTracker, triggering a change_count increase.
+    /// Note: If the service already exists, it will be updated instead.
+    pub fn add_service(&mut self, service: Service) {
+        if let Ok(_) = self.update_service(service.clone()) {
+            return;
+        }
+        self.change_count += 1;
+        let key = self.service_registry.add_service(service.service_id);
+        self.service_versions.insert(key, service.version);
+        self.service_list.push(ServiceRevision {
+            change_count: self.change_count,
+            service: key,
+        });
+    }
+
+    /// Update a `service` to a new revision, triggering a change_count increase.
+    ///
+    /// Returns an error if the `service` was never initialized/added.
+    pub fn update_service(&mut self, service: Service) -> Result<u32> {
+        let key = self.service_registry
+            .lookup_key(service.service_id)
+            .ok_or("Service not found")?;
+
+        if let Some(ver) = self.service_versions.get_mut(&key) {
+            if *ver == service.version {
+                return Ok(self.change_count);
+            }
+            *ver = service.version;
+        } else {
+            return Err("Service not found".into());
+        }
+
+        // Check to see if this service has been updated since initialization
+        let mut svc_index: Option<usize> = None;
+        for (i, svc) in self.service_list.iter().enumerate() {
+            if svc.service == key {
+                svc_index = Some(i);
+                break;
+            }
+        }
+        self.change_count += 1;
+        if let Some(svc_index) = svc_index {
+            let mut svc = self.service_list.remove(svc_index);
+            svc.change_count = self.change_count;
+            self.service_list.push(svc);
+        } else {
+            self.service_list.push(
+                ServiceRevision {
+                    change_count: self.change_count,
+                    service: key,
+                }
+            )
+        }
+        Ok(self.change_count)
+    }
+
+    /// Returns the new service versions since the provided `client_set`.
+    pub fn change_count_delta(&self, client_set: &mut ServiceList) -> Option<Vec<Service>> {
+        if self.change_count <= client_set.change_count {
+            return None;
+        }
+        let mut svc_iter = self.service_list.iter().rev();
+        let mut svc_delta = Vec::new();
+        while let Some(svc) = svc_iter.next() {
+            if svc.change_count <= client_set.change_count {
+                break;
+            }
+            if let Some(ver) = self.service_versions.get(&svc.service) {
+                if let Some(svc_id) = self.service_registry.lookup_id(svc.service) {
+                    svc_delta.push(Service {
+                        service_id: svc_id,
+                        version: (*ver).clone(),
```

Comment:
Could probably kill some of the clone()s in this module, but I don't care for now. This is the most important one

Reply:
These will need to be clone'd since they're going to have to be passed back out to serde at some point, which will exceed the scope of which a borrow would work from the caller.

Reply:
In the case here, removing clone() annoys the compiler cause *ver cannot be moved out of borrowed context.